backend/api/src/realms.rs

```diff
@@ -25,73 +38,126 @@ impl Realm {
+            .collect()
     }
 }
 
 pub(crate) struct Tree {
     pub(crate) realms: HashMap<u64, Realm>,
-    children: HashMap<u64, Vec<u64>>,
+    from_path: HashMap<String, u64>,
 }
 
 impl Tree {
-    pub(crate) async fn raw_from_db(
-        db: &Pool,
-    ) -> anyhow::Result<impl TryStream<Ok = Realm, Error = impl std::error::Error>> {
-        let row_stream = db.get().await?
-            .query_raw(
-                "select id, name, parent from realms",
-                std::iter::empty(),
-            ).await?;
-
-        Ok(row_stream.map_ok(|row| Realm {
-            key: row.get_key(0),
-            name: row.get(1),
-            parent_key: row.get_key(2),
-        }))
-    }
-
     pub(crate) async fn from_db(db: &Pool) -> anyhow::Result<Self> {
         // We store the nodes of the realm tree in a hash map
         // accessible by the database ID
-        let realms = Self::raw_from_db(db).await?
+        let mut realms = db.get().await?
+            .query_raw(
+                "select id, name, parent, path_segment from realms",
+                std::iter::empty(),
+            ).await?
+            .map_ok(|row| {
+                let key = row.get_key(0);
+                Realm {
+                    key,
+                    name: row.get(1),
+                    parent_key: if key == 0 { None } else { Some(row.get_key(2)) },
+                    path_segment: row.get(3),
+                    child_keys: vec![],
+                }
+            })
             .map_ok(|realm| (realm.key, realm))
             .try_collect::<HashMap<_, _>>().await?;
 
         // With this, and the `parent` member of the `Realm`,
         // we already have quick access to the data of a realm's parent.
-        // To also get to the children quickly we maintain another map.
-        let mut children = <HashMap<_, Vec<_>>>::new();
+        // To also get to the children quickly we maintain a corresponding list
+        // for each realm
+        let keys = realms.values()
+            .filter_map(
+                |realm| realm.parent_key.map(
+                    |parent_key| (realm.key, parent_key)
+                )
+            )
+            .collect::<Vec<_>>();
+        for (key, parent_key) in keys {
+            let parent = realms.get_mut(&parent_key).context("invalid parent")?;
+            parent.child_keys.push(key);
+        }
+
+        // After this point, we should know the tree structure to be valid.
+        // That is, we can now safely panic if we can't find things in our maps/lists;
+        // that's totally a bug in this code, then, not an inconsistency in the db.
+
+        // We also need a map from the full path to the proper realm,
+        // and conversely, we want to cache the full path inside the realm.
+        let from_path = index_by_path(&mut realms);
+
+        fn index_by_path(
+            realms: &mut HashMap<u64, Realm>,
+        ) -> HashMap<String, u64> {
```

Comment:
```suggestion
        fn index_by_path(realms: &mut HashMap<u64, Realm>) -> HashMap<String, u64> {
```

That is still comfortably below 100.

But thinking about this: why a function for this anyway? Seems like it can easily be written right in the outer function. The innermost recursive function can/has to stay of course.

Reply:
:shrug:

But this is moot anyway, since we build the map differently now.